### PR TITLE
Added comparable row-oriented representation of a collection of [`Array`].

### LINKED ISSUE #1280: Is there a module like arrow-rs's `row` in arrow2?
https://docs.rs/arrow/latest/arrow/row/index.html

Description: A comparable row-oriented representation of a collection of [Array](https://docs.rs/arrow/latest/arrow/array/trait.Array.html)

### PR DESCRIPTION
**Pull request by RinChanNOWWW**
Implemtation of [a comparable row format](https://docs.rs/arrow/latest/arrow/row/index.html) in arrow2.

The `row` module is used to convert columns into rows by `SortOption`s. After converting columns to rows, we can compare two 'row's directly without `build_compare`. This is very useful for `order by` execution in query engines and can help to improve the performance.

A `Rows` is fix-sized, and use `Box<[u8]>`  to store data. So there won't be too much heap allocation like `Vec<u8>`.

Struture:

- fixed: encode fix-sized types.
- variale: encode variable-sized types.
- dictionary: encode dictionary type.
- interner: dictionary type sort helper.

closes #1280

### CODE REVIEW

File: src/lib.rs
Comment (ritchie46) [inline src/lib.rs]: This has performance benefits.

We could add this behind a nightly feature (and maybe activate that feature in a build.rs file if a nightly compiler is detected).

We can write a custom trait to deal with the dispatch:

```rust
use std::hash::{BuildHasher, Hash};

pub trait HashSingle: BuildHasher {
    #[inline]
    fn hash_single<T: Hash>(&self, x: T) -> u64
    where
        Self: Sized,
    {
        #[cfg(feature = "nightly")]
        {
            self.hash_one(x)
        }
        #[cfg(not(feature = "nightly"))]
        {
            use std::hash::Hasher;
            let mut hasher = self.build_hasher();
            x.hash(&mut hasher);
            hasher.finish()
        }
    }
}

impl<T: BuildHasher> HashSingle for T {}

```

File: src/compute/sort/row/mod.rs
Comment (jorgecarleitao) [inline src/compute/sort/row/mod.rs:334]:
```suggestion

unsafe impl<'a> crate::trusted_len::TrustedLen for RowsIter<'a> {}
```

since this iterator is trusted len.

File: src/compute/sort/row/mod.rs
Comment (jorgecarleitao) [inline src/compute/sort/row/mod.rs:344]:
```suggestion
```

not used :)

File: src/compute/sort/row/mod.rs
Comment (jorgecarleitao) [inline src/compute/sort/row/mod.rs:611]: move to `tests/` (all tests to public APIs go there so that we verify that the functions and structs are available from the crate instead of `super`)

**Review by jorgecarleitao — APPROVED** [review]
This looks really great. Thanks a lot, @RinChanNOWWW !

I left minor comments that I do not think need an action - they are improvements to the API and performance optimizations that I thought about while going through it.

I will wait for the CI to be green and we can merge it 👍

File: src/compute/sort/row/mod.rs
Comment (jorgecarleitao) [reply src/compute/sort/row/mod.rs:237]: This is almost parallelizable - it is changing `rows`. However, there is still an optimization since modifying `rows` is `O(1)` but encoding is `O(C)`. Will continue to think about this.

**Comment by RinChanNOWWW** [comment]
I'm adding the row encoding for List types. And then I will apply the advices above.

File: src/compute/sort/row/mod.rs
Comment (RinChanNOWWW) [reply src/compute/sort/row/mod.rs:611]: How about some tests using private methods of a module?

Comment (RinChanNOWWW) [reply src/compute/sort/row/mod.rs:611]: Should we spilit the two types of tests?

File: src/compute/sort/row/mod.rs
Comment (RinChanNOWWW) [reply src/compute/sort/row/mod.rs:243]: I'm puzzled about how to convert this into a parallel one?

File: src/compute/sort/row/mod.rs
Comment (jorgecarleitao) [reply src/compute/sort/row/mod.rs:630]: exactly - tests to public functionality go to public, rest stays here :)

File: src/compute/sort/row/mod.rs
Comment (jorgecarleitao) [reply src/compute/sort/row/mod.rs:243]: I agree.

The idea would be that we can build `Vec<u8>` per column in parallel and only write them to the rows afterwards. However, I realized later that "encoding" is very simple and low overhead, so I am not sure it is worth.

File: Cargo.toml
Comment (jorgecarleitao) [inline Cargo.toml:36]:
```suggestion
hashbrown = { version = "0.12", default-features = false, optional = true }
```

together with `compute_sort = ["compute_take", "hashbrown"]`

so that we do not drag this dependency when we do not need it.

**Review by jorgecarleitao — APPROVED** [review]

File: src/compute/sort/row/interner.rs
Comment (alamb) [inline src/compute/sort/row/interner.rs:1]: FWIW this is quite similar (possibly copied directly from ) arrow-rs: https://github.com/apache/arrow-rs/blob/65d5576/arrow/src/row/interner.rs

Which is fine, I just wanted to give @tustvold credit where credit is due

Comment (RinChanNOWWW) [reply src/compute/sort/row/interner.rs:1]: Yes, the basic idea is the same as arrow-rs, and I adapted the codes to fit arrow2 data structures. Thanks for arrow-rs and @tustvold.

Comment (tustvold) [reply src/compute/sort/row/interner.rs:1]: FWIW I think it is pretty cool that the flow of technology is now from arrow to arrow2, it's a nice validation of the work I put into porting the ideas behind arrow2 across.Perhaps eventually we can combine efforts :smile:

File: build.rs
Comment (Xuanwo) [inline build.rs:5]: I prefer to introduce a new feature called `nightly` instead of adding a new dep in `build-dependencies`.

And I think the current implementation is wrong.

```diff
- #![cfg_attr(feature = "nightly_build", feature(build_hasher_simple_hash_one))]
+ #![cfg_attr(nightly_build, feature(build_hasher_simple_hash_one))]
```

Comment (Xuanwo) [reply build.rs:5]: We can take `hashbrown`'s implementation for a look: https://github.com/rust-lang/hashbrown/pull/292

File: src/compute/sort/row/interner.rs
Comment (Xuanwo) [inline src/compute/sort/row/interner.rs:23]: To my best knowledge, rust std's `HashMap` uses the same implementation as `hashbrown`. Is there any reason to hashbrown directly?

File: build.rs
Comment (RinChanNOWWW) [reply build.rs:5]: > I prefer to introduce a new feature called nightly instead of adding a new dep in build-dependencies.

But users need to add this to the features list. `build.rs` can help to select automatically.

File: src/compute/sort/row/interner.rs
Comment (RinChanNOWWW) [reply src/compute/sort/row/interner.rs:23]: > Is there any reason to hashbrown directly?

No, it just follows arrow-rs.

File: build.rs
Comment (Xuanwo) [reply build.rs:5]: > But users need to add this to the features list. `build.rs` can help to select automatically.

I prefer adding a new feature instead. Leave this for @jorgecarleitao to decide.

File: src/compute/sort/row/interner.rs
Comment (Xuanwo) [reply src/compute/sort/row/interner.rs:23]: How about using std's `HashMap` instead? So we don't need to introduce `hashbrown`.

Comment (tustvold) [reply src/compute/sort/row/interner.rs:23]: At least historically the raw entry API, which this requires, isn't exposed in std

Comment (Xuanwo) [reply src/compute/sort/row/interner.rs:23]: > At least historically the raw entry API, which this requires to store the values separately from the HashMap, isn't exposed in std

Thanks for explanation!

Is the raw entry API we use here something different from https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html?

Comment (Xuanwo) [reply src/compute/sort/row/interner.rs:23]: By the way, we have https://doc.rust-lang.org/std/collections/hash_map/enum.RawEntryMut.html under nightly feature `hash_raw_entry`.

File: src/compute/sort/row/mod.rs
Comment (Xuanwo) [reply src/compute/sort/row/mod.rs:320]: Lesson learned.

File: build.rs
Comment (jorgecarleitao) [reply build.rs:5]: I would also prefer a specific feature, but both are mergable :) - we use this for `simd` also where users activate simd if they are in nightly. Maybe the nightly feature can activate both simd and this path? I can work out the details once we merge this :)

File: src/compute/sort/row/interner.rs
Comment (jorgecarleitao) [reply src/compute/sort/row/interner.rs:1]: > FWIW I think it is pretty cool that the flow of technology is now from arrow to arrow2

I believe it has always been both ways - which is part of the beauty of open source - people can do (almost) whatever they want without having to be worried about who "owns" what.

**Comment by Xuanwo** [comment]
For the latest CI failing in `Check and test / Clippy (pull_request)`: https://github.com/jorgecarleitao/arrow2/actions/runs/3390608894/jobs/5636710157

```shell
error: the borrowed expression implements the required traits
   --> src/io/parquet/write/pages.rs:301:48
    |
301 |         let boolean = BooleanArray::from_slice(&[false, false, true, true]).boxed();
    |                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: change this to: `[false, false, true, true]`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrow
```

Welcome to rust 1.65! We can fix them by `cargo clippy --fix`.

**Comment by RinChanNOWWW** [comment]
62 files have clippy problems under rust 1.65 (83 files under rust 1.67). We can fix them in another pr.

**Comment by jorgecarleitao** [comment]
Thank you @tustvold for the contribution (eagerly await the blog post!), @RinChanNOWWW for the port, and @Xuanwo and @ritchie46 for reviewing it! 🙇

File: src/compute/sort/row/interner.rs
Comment (ritchie46) [reply src/compute/sort/row/interner.rs:1]: Awesome work on thus @tustvold!

**Comment by alamb** [comment]
> Thank you @tustvold for the contribution (eagerly await the blog post!),

In case anyone missed it, here is the blog post about the format

https://arrow.apache.org/blog/2022/11/07/multi-column-sorts-in-arrow-rust-part-1/
https://arrow.apache.org/blog/2022/11/07/multi-column-sorts-in-arrow-rust-part-2/

**Comment by RinChanNOWWW** [comment]
image

> and is then followed by the unpadded length of this final block as a single byte in place of a continuation token

The demo in https://arrow.apache.org/blog/2022/11/07/multi-column-sorts-in-arrow-rust-part-2/ seems be wrong.

**Comment by alamb** [comment]
Thanks for the report @RinChanNOWWW  -- there is a PR to correct this https://github.com/apache/arrow-site/pull/270

**Pull request merged**
